New Feature: Database Partitions

davisp opened the pull request:
## Overview

This PR introduces a new feature, user-defined partitioned databases.

A new kind of database can be created with the ?partitioned=true option. All documents within the database must have document ids of the following format;

partition_name:doc_id

both partition_name and doc_id must follow the couchdb id format (can't begin with _, etc).

All documents with the same partition_name are guaranteed to be mapped to the same shard range. When querying an index, the new /db/_partition/$partition/_view endpoint can query the view more efficiently, by only consulting the single shard range holding the partition. This is much more efficient and scales the same way that primary key lookup (GET /dbname/docid) does (approximately linearly).

## Testing recommendations

The PR contains multiple tests for basic functionality and all existing tests still pass. When testing the PR, it is important to try the feature yourself, interactively, with docs and views of your choosing, to give us confidence in this new feature.

## Related Issues or Pull Requests

This supersedes PR #1605.

## Checklist

- [x] Code is written and works correctly;
- [x] Changes are covered by tests;
- [ ] Documentation reflects the changes;

davisp commented on src/chttpd/src/chttpd_db.erl:
```diff
@@ -253,6 +255,41 @@ handle_view_cleanup_req(Req, Db) ->
+            },
+            case Rest of
+                [] ->
+                    do_db_req(NewReq, Db);
```
That's actually dead code. `Rest == []` will have been caught by the first function clause. Must've from before the refactor. Will tidy this up.

davisp replied:
Fixed by removing the case and just doing `hd(Rest)`.

garrensmith commented on src/couch_mrview/src/couch_mrview_index.erl:
```diff
@@ -264,6 +267,29 @@ get_ddoc(DbName, DesignDocs, DDocId) ->
+        design_opts = DesignOpts
+    } = State,
+    DbPartitioned = couch_db:is_partitioned(Db),
+    ViewPartitioned = proplists:get_value(<<"partitioned">>, DesignOpts),
```
I think we should do `ViewPartitioned = proplists:get_value(<<"partitioned">>, DesignOpts, DbPartitioned),` here. That would reduce two of the case statement lines since if there is no option then the view partition option is the same as the `DbPartitioned` value.

garrensmith replied:
Also when should we use `couch_util:get_value` and when should we use `proplists:get_value`?

davisp replied:
Fixed. No idea why my brain when to proplists there. I try and use couch_util:get_value exclusively.

garrensmith commented on test/elixir/test/partition_view_test.exs:
```diff
@@ -0,0 +1,408 @@
+        }
+      )
+
+    Logger.debug("#{inspect(resp)}")
```
We can remove this

davisp replied:
Fixed.

iilyak commented on src/couch/src/couch_partition.erl:
```diff
@@ -0,0 +1,143 @@
+
+hash(<<"_", _/binary>> = DocId) ->
+    erlang:crc32(DocId);
+
```
Is extra new line intentional?

davisp replied:
Fixed.

iilyak commented on src/fabric/src/fabric_util.erl:
```diff
@@ -302,7 +304,17 @@ is_users_db(DbName) ->
     Suffix =:= couch_db:dbname_suffix(Path).
 
-fake_db(DbName, Opts) ->
+make_cluster_db(#shard{dbname = DbName, opts = Options}) ->
```
The name is confusing. `make` is too close to `create` in meaning.

davisp replied:
You mean that it could be confused with `fabric:create_db/1` in that it might be actually creating shard maps and shard files on disk instead of creating a database handle? How about `open_clustered_db`?

davisp replied:
Fixed with open_clustered_db.

iilyak commented:
I didn't finish my review yet. But this is looking great so far.
There is one aspect of the codebase which I don't like. However it is very easy to fix. 
There are 3 places where we leak implementation detail that the partition separator is ":". 
Would you mind adding following two functions into `couch_partition.erl`:
```
-compile({inline, [
    start_key/1,
    end_key/1
]}).
start_key(Partition) -> 
    <<Partition/binary, ":">>.

end_key(Partition) -> 
    <<Partition/binary, ";">>.
```

Then we can use these functions:
- in [couch_bt_engine:get_partition_info/2](https://github.com/apache/couchdb/pull/1789/files#diff-eaad4742764a0f11879a4bb6da73f47dR305)
- in [couch_mrview_util:apply_all_docs_partition/2](https://github.com/apache/couchdb/pull/1789/files#diff-2bd31f00dcfba780c6c65bc69c256854R729) instead of `PARTITION_START` and `PARTITION_END` macro.

The third place I mentioned is in [couch_mrview_updater:partition/1](https://github.com/apache/couchdb/pull/1789/files#diff-0a95c04f539e3b4a470eaa7339123ff5R405). I already have a comment about it.

iilyak commented:
Some tests need styling updates:
```
** (Mix) mix format failed due to --check-formatted.
The following files were not formatted:

  * test/partition_view_test.exs
  * test/partition_mango_test.exs
  * test/partition_crud_test.exs
  * test/all_docs_test.exs
```

iilyak commented:
After updating Makefile to skip elixir-check-formatted I am getting multiple test failures in Elixir test suite. Some of them might not be related:
```
ReplicationTest
  * test unauthorized replication cancellation (1038.8ms)

  1) test unauthorized replication cancellation (ReplicationTest)
     test/replication_test.exs:156
     Expected truthy, got false
     code: assert HTTPotion.Response.success?(resp)
     stacktrace:
       test/replication_test.exs:2: anonymous fn/1 in ReplicationTest.set_user_context/1
       (ex_unit) lib/ex_unit/on_exit_handler.ex:140: ExUnit.OnExitHandler.exec_callback/1
       (ex_unit) lib/ex_unit/on_exit_handler.ex:126: ExUnit.OnExitHandler.on_exit_runner_loop/0
```

```
  * test non-admin or reader user on source - remote-to-remote (90.7ms)

  2) test non-admin or reader user on source - remote-to-remote (ReplicationTest)
     test/replication_test.exs:248
     Expected truthy, got false
     code: assert HTTPotion.Response.success?(resp)
     stacktrace:
       test/replication_test.exs:2: ReplicationTest.create_user/1
       test/replication_test.exs:2: ReplicationTest.set_user_context/1
       (elixir) lib/enum.ex:1899: Enum."-reduce/3-lists^foldl/2-0-"/3
       test/replication_test.exs:2: ReplicationTest.__ex_unit_setup_0/1
       test/replication_test.exs:1: ReplicationTest.__ex_unit__/2
```

```
  * test non-admin user on target - remote-to-local (51.4ms)

  3) test non-admin user on target - remote-to-local (ReplicationTest)
     test/replication_test.exs:243
     Expected truthy, got false
     code: assert HTTPotion.Response.success?(resp)
     stacktrace:
       test/replication_test.exs:2: ReplicationTest.create_user/1
       test/replication_test.exs:2: ReplicationTest.set_user_context/1
       (elixir) lib/enum.ex:1899: Enum."-reduce/3-lists^foldl/2-0-"/3
       test/replication_test.exs:2: ReplicationTest.__ex_unit_setup_0/1
       test/replication_test.exs:1: ReplicationTest.__ex_unit__/2
```

```
  * test non-admin user on target - local-to-local (64.0ms)

  4) test non-admin user on target - local-to-local (ReplicationTest)
     test/replication_test.exs:243
     Expected truthy, got false
     code: assert HTTPotion.Response.success?(resp)
     stacktrace:
       test/replication_test.exs:2: ReplicationTest.create_user/1
       test/replication_test.exs:2: ReplicationTest.set_user_context/1
       (elixir) lib/enum.ex:1899: Enum."-reduce/3-lists^foldl/2-0-"/3
       test/replication_test.exs:2: ReplicationTest.__ex_unit_setup_0/1
       test/replication_test.exs:1: ReplicationTest.__ex_unit__/2
```

```

  * test non-admin user on target - local-to-remote (36.3ms)

  5) test non-admin user on target - local-to-remote (ReplicationTest)
     test/replication_test.exs:243
     Expected truthy, got false
     code: assert HTTPotion.Response.success?(resp)
     stacktrace:
       test/replication_test.exs:2: ReplicationTest.create_user/1
       test/replication_test.exs:2: ReplicationTest.set_user_context/1
       (elixir) lib/enum.ex:1899: Enum."-reduce/3-lists^foldl/2-0-"/3
       test/replication_test.exs:2: ReplicationTest.__ex_unit_setup_0/1
       test/replication_test.exs:1: ReplicationTest.__ex_unit__/2
```

```
  * test non-admin or reader user on source - local-to-remote (71.0ms)

  6) test non-admin or reader user on source - local-to-remote (ReplicationTest)
     test/replication_test.exs:248
     Expected truthy, got false
     code: assert HTTPotion.Response.success?(resp)
     stacktrace:
       test/replication_test.exs:2: ReplicationTest.create_user/1
       test/replication_test.exs:2: ReplicationTest.set_user_context/1
       (elixir) lib/enum.ex:1899: Enum."-reduce/3-lists^foldl/2-0-"/3
       test/replication_test.exs:2: ReplicationTest.__ex_unit_setup_0/1
       test/replication_test.exs:1: ReplicationTest.__ex_unit__/2

```

```
  * test non-admin or reader user on source - local-to-local (49.4ms)

  7) test non-admin or reader user on source - local-to-local (ReplicationTest)
     test/replication_test.exs:248
     Expected truthy, got false
     code: assert HTTPotion.Response.success?(resp)
     stacktrace:
       test/replication_test.exs:2: ReplicationTest.create_user/1
       test/replication_test.exs:2: ReplicationTest.set_user_context/1
       (elixir) lib/enum.ex:1899: Enum."-reduce/3-lists^foldl/2-0-"/3
       test/replication_test.exs:2: ReplicationTest.__ex_unit_setup_0/1
       test/replication_test.exs:1: ReplicationTest.__ex_unit__/2
```

```
  * test non-admin or reader user on source - remote-to-local (192.6ms)

  8) test non-admin or reader user on source - remote-to-local (ReplicationTest)
     test/replication_test.exs:248
     Expected truthy, got false
     code: assert HTTPotion.Response.success?(resp)
     stacktrace:
       test/replication_test.exs:2: ReplicationTest.create_user/1
       test/replication_test.exs:2: ReplicationTest.set_user_context/1
       (elixir) lib/enum.ex:1899: Enum."-reduce/3-lists^foldl/2-0-"/3
       test/replication_test.exs:2: ReplicationTest.__ex_unit_setup_0/1
       test/replication_test.exs:1: ReplicationTest.__ex_unit__/2
```

```
  * test non-admin user on target - remote-to-remote (54.9ms)

  9) test non-admin user on target - remote-to-remote (ReplicationTest)
     test/replication_test.exs:243
     Expected truthy, got false
     code: assert HTTPotion.Response.success?(resp)
     stacktrace:
       test/replication_test.exs:2: ReplicationTest.create_user/1
       test/replication_test.exs:2: ReplicationTest.set_user_context/1
       (elixir) lib/enum.ex:1899: Enum."-reduce/3-lists^foldl/2-0-"/3
       test/replication_test.exs:2: ReplicationTest.__ex_unit_setup_0/1
       test/replication_test.exs:1: ReplicationTest.__ex_unit__/2
```

```
PartitionCrudTest
  ...
  * test saves attachment with partitioned doc (171.8ms)

 10) test saves attachment with partitioned doc (PartitionCrudTest)
     test/partition_crud_test.exs:246
     Assertion with == failed
     code:  assert body["_attachments"] == %{"foo.txt" => %{"content_type" => "text/plain", "digest" => "md5-OW2BoZAtMqs1E+fAnLpNBw==", "length" => 31, "revpos" => 1, "stub" => true}}
     left:  %{"foo.txt" => %{"content_type" => "text/plain", "length" => 31, "revpos" => 1, "stub" => true, "digest" => "md5-wJORG+sY7HTKqCWOnk5tIQ=="}}
     right: %{"foo.txt" => %{"content_type" => "text/plain", "length" => 31, "revpos" => 1, "stub" => true, "digest" => "md5-OW2BoZAtMqs1E+fAnLpNBw=="}}
     stacktrace:
       test/partition_crud_test.exs:269: (test)
```

```
BasicsTest
  ...
  * test PUT error when body not an object (147.4ms)

 11) test PUT error when body not an object (BasicsTest)
     test/basics_test.exs:243
     Assertion with == failed
     code:  assert resp.status_code() == 400
     left:  500
     right: 400
     stacktrace:
       test/basics_test.exs:246: (test)
```

davisp commented:
@iilyak Those inline calls I believe only apply to the module, I don't think that'll affect anywhere else they're used as external functions? However we could at least hoist the macros into a couch_partitions.hrl for re-use.

For the mix format, those should be fixed after one of my recent-ish force pushes. I rebased for something else and forgot to check that after we added Credo.

I'm looking into the other test failures to see what's going on with those.

jaydoane commented:
After `make clean && git clean -xffd && ./configure --dev && make`, I'm seeing numerous setup failures and a couple test failures running `make eunit apps=chttpd`:
```
=======================================================
  Failed: 2.  Skipped: 0.  Passed: 184.
```
compared to a successful result when run against the master commit this is based against:
```
=======================================================
  All 233 tests passed.
```
The two failures are
```
chttpd_view:123: t_check_include_docs_throw_validation_error...*failed*
in function chttpd_view:'-t_check_include_docs_throw_validation_error/0-fun-0-'/0 (src/chttpd_view.erl, line 127)
**error:{assertException,
    [{module,chttpd_view},
     {line,127},
     {expression,
         "multi_query_view ( Req , db , ddoc , << \"v\" >> , [ Query ] )"},
     {pattern,"{ throw , Throw , [...] }"},
     {unexpected_exception,
         {error,function_clause,
             [{couch_db,is_clustered,[db],[{file,...},{...}]},
              {fabric_util,validate_args,3,[{...}|...]},
              {lists,map,2,[...]},
              {chttpd_view,multi_query_view,5,...},
              {chttpd_view,
                  '-t_check_include_docs_throw_validation_error/0-fun-0-',...},
              {eunit_test,...},
              {...}|...]}}]}
  output:<<"">>

chttpd_view:132: t_check_user_can_override_individual_query_type...*failed*
in function couch_db:is_clustered/1 (src/couch_db.erl, line 217)
  called as is_clustered(db)
```

davisp commented:
@jaydoane chttpd tests should be fixed now. Am running through checking that everything else works as well.

Fix for the ones you noted was to use `test_util:fake_db/1` instead of the atom `db`.

davisp commented:
@jaydoane @iilyak All tests both eunit and elixir should be passing now. Apologies for the delay!

davisp commented:
@iilyak I've created helper functions for the start/end key leak and updated couch_mrview_updater to use couch_partition:extract to also avoid the separator leak there.

@jaydoane I've updated the check in couch_partition to avoid the config calls when a database is not partitioned.

iilyak commented on src/couch/src/couch_partition.erl:
```diff
@@ -0,0 +1,156 @@
+
+    if not IsPartitioned -> ok; true ->
+
+        DbsDbName = config:get("mem3", "shards_db", "_dbs"),
```
It is either github loosing comments or I forgot to comment about it. I think the [couch_db:is_system_db_name/1](https://github.com/apache/couchdb/blob/master/src/couch/src/couch_db.erl#L1765) need to be updated to respect configured suffices.

davisp replied:
Not entirely sure though that's a different PR if so. The places its used I think make it look like not but I dunno as I've not used it in this PR.

davisp commented:
Nope hadn't seen those. Will address them today.

davisp commented on src/couch/src/couch_bt_engine_header.erl:
```diff
@@ -69,7 +69,8 @@
     compacted_seq,
-    purge_infos_limit = 1000
+    purge_infos_limit = 1000,
+    props_ptr
```
Oops I forgot the resolve button doesn't post the comment. The downgrade logic was already merged: https://github.com/apache/couchdb/commit/854f3d9fcbd8db91ab208e27038c252d91841fcf

davisp commented on src/couch/src/couch_partition.erl:
```diff
@@ -0,0 +1,142 @@
+    IsSysDb = lists:member(DbName, SysDbNames)
+            orelse lists:member(Suffix, Suffices),
+
+    if not (IsSysDb and IsPartitioned) -> ok; true ->
```
Addressed this by gating on partitioned before doing the further config checks.

davisp commented on src/couch/src/couch_db.erl:
```diff
@@ -583,6 +603,10 @@ get_db_info(Db) ->
         Else1 -> Else1
     end,
+    Props = case couch_db_engine:get_props(Db) of
+        undefined -> null;
```
Good catch. I've fixed this and will add a test for it. Unless I'm mistaken the general project guidance on client libraries is to not blow up as JSON keys are added to responses.

rnewson replied:
libraries, if they're sensible, should not be broken if new keys appear. Where we fall down sometimes is that libraries make assumptions about the types of the value (update_seq being an integer and then a string, but was json all along).

imo, if we add a new key and a library breaks, then the library is at fault.

davisp commented on src/couch/src/couch_db.erl:
```diff
@@ -604,12 +628,21 @@ get_db_info(Db) ->
-    DbName = mem3:dbname(ShardDbName),
+
+get_partition_info(#db{} = Db, Partition) when is_binary(Partition) ->
+    Sizes = couch_db_engine:get_partition_info(Db, Partition),
```
Done.

davisp commented on src/couch/src/couch_db.erl:
```diff
@@ -604,12 +628,21 @@ get_db_info(Db) ->
+
+
+get_design_docs(#db{name = <<"shards/", _:18/binary, DbFullName/binary>>}) ->
+    DbName = ?l2b(filename:rootname(filename:basename(?b2l(DbFullName)))),
```
This was out of date. It got swapped to `mem3:dbname/1` cause this was wrong to begin with.

davisp commented on src/fabric/src/fabric.erl:
```diff
@@ -378,25 +392,26 @@ query_view(Db, Options, GroupId, ViewName, Callback, Acc0, QueryArgs)
-    Db = dbname(DbName), View = name(ViewName),
-    case fabric_util:is_users_db(Db) of
+    query_view(Db, Options, DDoc, ViewName, Callback, Acc0, QueryArgs);
+query_view(Db342, Options, DDoc, ViewName, Callback, Acc0, QueryArgs0) ->
```
Heh, forgot to rename that. Was tired and did that to make sure I didn't re-use a variable name. Reverted back to Db.

davisp commented on src/fabric/src/fabric_db_partition_info.erl:
```diff
@@ -0,0 +1,98 @@
+
+% for JS to work nicely we need to convert the size list
+% to a jiffy object
+format_partition(PartitionInfo) ->
```
`GET /dbname/_partition/name` returns the partition info blob as JSON.

iilyak replied:
I see

davisp commented on src/fabric/src/fabric_view.erl:
```diff
@@ -309,10 +317,23 @@ index_of(X, [X|_Rest], I) ->
+        throw({bad_request, <<"partition specified on non-partitioned db">>})
+    end,
+    DbName = fabric:dbname(Db),
+    case {Args#mrargs.stable, Partition} of
```
I think you missed the subtle `mem3:shards` vs `mem3:ushards` there. I'll add a comment to make that clearer.

davisp replied:
Let me know if it was different though. I won't resolve this till then as that particular bit of code was a bit gnarly for me the first time I read it.

iilyak replied:
> I think you missed the subtle mem3:shards vs mem3:ushards there. I'll add a comment to make that clearer.

Indeed I missed that.

davisp commented on src/mango/src/mango_idx.erl:
```diff
@@ -329,6 +339,87 @@ gen_name(Idx, Opts0) ->
+    case proplists:get_value(partitioned, Opts) of
+        B when is_boolean(B) ->
+            B;
+        default ->
```
Ah, I could see that being confusing. I'll update the atom to be `db_default` rather and add a comment indicating that it means to default to the partition setting on the database.

davisp commented on src/mango/src/mango_idx.erl:
```diff
@@ -329,6 +339,87 @@ gen_name(Idx, Opts0) ->
+        B when is_boolean(B) ->
+            B;
+        default ->
+            undefined
```
The ternary logic is due to the "when unspecified it defaults to whatever the database is configured to use" which isn't plumbed through everywhere. It certainly adds a bit of complexity to the implementation although from the user's perspective it makes the most sense. If you create a partitioned database, one would expect to have partitioned indexes by default as that's the point of partitioning in the first place.

However, we also allow for non-partitioned (or global) indexes on partitioned databases as well as reject attempts to create partitioned indexes on non-partitioned databases.

davisp commented on src/mango/src/mango_idx.erl:
```diff
@@ -329,6 +339,87 @@ gen_name(Idx, Opts0) ->
+filter_partition_indexes(Indexes, Opts) ->
+    PFilt = case couch_util:get_value(partition, Opts) of
+        <<>> ->
+            fun(#idx{partitioned = P}) -> not P end;
```
I think it'd be better to let it crash in that case cause if partitioned isn't a boolean then its a bug that needs fixing?

iilyak replied:
Agree.

davisp commented on src/mango/src/mango_idx.erl:
```diff
@@ -329,6 +339,87 @@ gen_name(Idx, Opts0) ->
+    #doc{
+        body = {BodyProps}
+    } = DDoc,
+    NewProps = case couch_util:get_value(<<"options">>, BodyProps) of
```
BodyProps is user provided input so there's no guarantee that the `partitioned` key was provided in the JSON blob.

iilyak replied:
I made a typo in my example it should be as follows and I still think it would work:
```
{Opts} = couch_util:get_value(<<"options">>, BodyProps, {[]}),
Opt = {<<"partitioned">>, Partitioned},
New = lists:keystore(<<"partitioned">>, 1, Opts, Opt),
NewProps = lists:keystore(<<"options">>, 1, BodyProps, {<<"options">>, New}),
DDoc#doc{body = {NewProps}}.
```

- If `options` key is not provided the Opts would be `{[]}` since it is specified in default argument in the call to `couch_util:get_value/3`.
- If `partitioned` key is not provided it we would add it in lists:keystore/4.
- If `partitioned` key is provided in the `BodyProps` we inject it into `options->partitioned`
- If value for `partitioned` is not in the right format is handled elsewhere.

davisp commented on src/mango/src/mango_opts.erl:
```diff
@@ -296,6 +309,23 @@ validate_fields(Value) ->
+validate_partition(<<>>) ->
+    {ok, <<>>};
+validate_partition(Partition) ->
+    couch_partition:validate_partition(Partition),
```
It'll get caught by the normal error handling which has clauses for it but we could wrap/unwrap it at the mango level for consistency if you like.

iilyak replied:
Not a big deal

davisp commented on src/mango/src/mango_cursor_text.erl:
```diff
@@ -93,6 +94,7 @@ execute(Cursor, UserFun, UserAcc) ->
     } = Cursor,
     QueryArgs = #index_query_args{
         q = mango_selector_text:convert(Selector),
+        partition = get_partition(Opts, nil),
```
The `null` in mango_cursor_view.erl is for when its presented in the JSON explain output.

iilyak replied:
I see

davisp commented on src/couch_mrview/src/couch_mrview_util.erl:
```diff
@@ -476,6 +498,49 @@ fold_reduce({NthRed, Lang, View}, Fun,  Acc, Options) ->
+
+    case {DbPartitioned, Partition} of
+        {false, <<_/binary>>} ->
+            mrverror(<<"`partition` paramter is not support on this db">>);
```
Good catch. Updated.

davisp commented on src/couch/src/couch_db.erl:
```diff
@@ -184,11 +189,22 @@ reopen(#db{} = Db) ->
-clustered_db(DbName, UserCtx) ->
-    clustered_db(DbName, UserCtx, []).
+clustered_db(DbName, Options) when is_list(Options) ->
+    UserCtx = couch_util:get_value(user_ctx, Options),
```
Good idea. Have added that.

davisp commented on test/elixir/test/partition_view_test.exs:
```diff
@@ -0,0 +1,431 @@
+
+    # Avoid race conditions by attempting to get a full response
+    # from every shard before we do our update:false test
+    for i <- 1..12 do
```
Fixed.

davisp commented on src/couch/src/couch_db.erl:
```diff
@@ -604,12 +628,21 @@ get_db_info(Db) ->
+    throw({bad_request, <<"`partition` is not valid">>}).
+
+
+get_design_docs(#db{name = <<"shards/", _:18/binary, DbFullName/binary>>}) ->
```
I ended up just removing this with the change below to use `mem3:dbname/1`. This is just now keyed in `<<"shards/", _binary>>`.

davisp commented on src/couch_mrview/src/couch_mrview_util.erl:
```diff
@@ -598,6 +663,12 @@ validate_args(Args) ->
 
+    case get_extra(Args, partition) of
+        undefined -> ok;
+        Partition when is_binary(Partition) -> ok;
```
Good idea. Have added it.

davisp commented on test/elixir/test/partition_design_docs_test.exs:
```diff
@@ -0,0 +1,16 @@
+
+    url = "/#{db_name}/_partition/fake-key/_design/mrtest/"
+    resp = Couch.get(url)
+    assert resp.status_code == 404
```
Hrm. Mebbe. Its just showing that we can't hit that route via the _partition endpoint. Also I tend to not assert on human readable text as that gets brittle if things are revised for clarity. I'd be happy to add the error value assertion I guess but if 404 wasn't `"error":"not_found"` we've got much bigger problems.

iilyak replied:
ok

iilyak commented on src/fabric/src/fabric_view.erl:
```diff
@@ -309,10 +317,26 @@ index_of(X, [X|_Rest], I) ->
+        {true, undefined} ->
+            mem3:ushards(DbName);
+        {true, Partition} ->
+            mem3:ushards(DbName, <<Partition/binary, ":foo">>);
```
Can we add `couch_partition:range_key/1` or `partition_key/1`
```
range_key(Partition) ->
   <<Partition/binary, ":foo">>.
```

davisp replied:
Good idea. Went with shard_key for the name.

iilyak commented on src/mango/src/mango_error.erl:
```diff
@@ -104,6 +104,13 @@ info(mango_idx, {invalid_index_type, BadType}) ->
         <<"invalid_index">>,
         fmt("Invalid type for index: ~s", [BadType])
     };
+info(mango_idx, {partitoned_option_mismatch, BadDDoc}) ->
```
Still needs fixing.

davisp replied:
Huh. Didn't see this originally. Fixing now.

iilyak commented on src/fabric/src/fabric_view.erl:
```diff
@@ -309,10 +317,26 @@ index_of(X, [X|_Rest], I) ->
+        {false, undefined} ->
+            mem3:shards(DbName);
+        {false, Partition} ->
+            mem3:shards(DbName, <<Partition/binary, ":foo">>)
```
Can we add couch_partition:range_key/1 or partition_key/1?

jaydoane commented on src/couch_mrview/src/couch_mrview_util.erl:
```diff
@@ -476,6 +496,49 @@ fold_reduce({NthRed, Lang, View}, Fun,  Acc, Options) ->
+
+    case {DbPartitioned, Partition} of
+        {false, <<_/binary>>} ->
+            mrverror(<<"`partition` paramter is not supported on this db">>);
```
"paramter" is also a typo

davisp replied:
Fixed.

iilyak commented:
mango tests are failing on Trsavis

iilyak commented on src/mango/src/mango_idx.erl:
```diff
@@ -329,6 +339,87 @@ gen_name(Idx, Opts0) ->
+    DDoc#doc{body = {NewProps}}.
+
+
+set_idx_partitioned(Db, DDocProps) ->
```
bump

davisp replied:
Fixed.

davisp commented:
Yeah, those mango queries are making assertions on the hash value of a mango index definition which is oddly internals specific when its just wanting to make sure the right index was picked. I tried the cheap way of updating the hash values but looks like I'm gonna have to name them.

davisp commented:
Definitely gonna wait for a second +1 on such a big feature.

tonysun83 commented:
Running through the elixir tests I'm getting a 500's due to a `couch_ejson_compare:less_erl/2` function_clause. `less_erl` doesn't match with tuples. I wonder if I have the latest commits for this branch?

```
[error] 2019-01-02T06:24:38.655017Z node3@127.0.0.1 <0.16144.0> 7b878638a9 rexi_server: from: node1@127.0.0.1(<27635.14553.0>) mfa: fabric_rpc:map_view/5 throw:{function_clause,[{couch_ejson_compare,less_erl,[{<<"foo">>,<<"98">>},{<<"foo">>,<<"96">>}],[{file,"src/couch_ejson_compare.erl"},{line,54}]},{couch_ejson_compare,less_list,2,[{file,"src/couch_ejson_compare.erl"},{line,106}]},{couch_ejson_compare,less_json_ids,2,[{file,"src/couch_ejson_compare.erl"},{line,39}]},{lists,sort,2,[{file,"lists.erl"},{line,967}]},{couch_btree,query_modify,4,[{file,"src/couch_btree.erl"},{line,260}]},{couch_btree,add_remove,3,[{file,"src/couch_btree.erl"},{line,239}]},{couch_mrview_updater,'-write_kvs/6-fun-1-',8,[{file,"src/couch_mrview_updater.erl"},{line,347}]},{lists,zipwith,3,[{file,"lists.erl"},{line,449}]}]} [{couch_mrview_util,get_view_index_state,5,[{file,"src/couch_mrview_util.erl"},{line,142}]},{couch_mrview_util,get_view,4,[{file,"src/couch_mrview_util.erl"},{line,86}]},{couch_mrview,query_view,6,[{file,"src/couch_mrview.erl"},{line,265}]},{rexi_server,init_p,3,[{file,"src/rexi_server.erl"},{line,140}]}]
[error] 2019-01-02T06:24:38.654986Z node1@127.0.0.1 <0.14553.0> 7b878638a9 req_err(3510611728) badmatch : {error,
    {function_clause,
        [{couch_ejson_compare,less_erl,
             [{<<"foo">>,<<"98">>},{<<"foo">>,<<"96">>}],
             [{file,"src/couch_ejson_compare.erl"},{line,54}]},
         {couch_ejson_compare,less_list,2,
             [{file,"src/couch_ejson_compare.erl"},{line,106}]},
         {couch_ejson_compare,less_json_ids,2,
             [{file,"src/couch_ejson_compare.erl"},{line,39}]},
         {lists,sort,2,[{file,"lists.erl"},{line,967}]},
         {couch_btree,query_modify,4,
             [{file,"src/couch_btree.erl"},{line,260}]},
         {couch_btree,add_remove,3,[{file,"src/couch_btree.erl"},{line,239}]},
         {couch_mrview_updater,'-write_kvs/6-fun-1-',8,
             [{file,"src/couch_mrview_updater.erl"},{line,347}]},
         {lists,zipwith,3,[{file,"lists.erl"},{line,449}]}],
        [{couch_mrview_util,get_view_index_state,5,
             [{file,"src/couch_mrview_util.erl"},{line,142}]},
         {couch_mrview_util,get_view,4,
             [{file,"src/couch_mrview_util.erl"},{line,86}]},
         {couch_mrview,query_view,6,
             [{file,"src/couch_mrview.erl"},{line,265}]},
         {rexi_server,init_p,3,[{file,"src/rexi_server.erl"},{line,140}]}]}}
    [<<"mango_httpd:handle_find_req/2 L189">>,<<"mango_httpd:handle_req/2 L38">>,<<"chttpd:handle_req_after_auth/2 L320">>,<<"chttpd:process_request/1 L303">>,<<"chttpd:handle_request_int/1 L243">>,<<"mochiweb_http:headers/6 L128">>,<<"proc_lib:init_p_do_apply/3 L237">>]
```

davisp commented:
Latest version of this branch is 60bbe3f. CI passed so I'm wagering you're on an old version.

davisp commented:
Except now that I look we're apparently not running Elixir tests in make check yet which I must've just remembered wrongly as having been enabled. Am now trying to reproduce locally.

davisp commented:
Found it. Silly bug when I switched couch_mrview_updater to reuse a function in couch_partition that returned a slightly different result. Have forced pushed the fix.

eiri commented on src/chttpd/src/chttpd_db.erl:
```diff
@@ -254,6 +255,36 @@ handle_view_cleanup_req(Req, Db) ->
+        false ->
+            throw({bad_request, <<"database is not partitioned">>})
+    end.
+
```
We probably want to add a base clause here, like:
```erlang
handle_partition_req(Req, Db) ->
    db_req(Req, Db).
```

That'd give a simple 404 on `/#{db}/_partition` instead of the dreadful `function_clause`, `unknown_error` we have there now.

davisp replied:
I just added a 404 handler directly rather than falling through to db_req which might do something weird.

jaydoane commented on test/elixir/test/partition_view_update_test.exs:
```diff
@@ -0,0 +1,73 @@
+defmodule PartitionSizeTest do
```
I believe this should be `PartitionViewUpdateTest` so it doesn't conflict with another `PartitionSizeTest` in partition_size_test.exs:
```
== Compilation error in file ../src/couchdb/test/elixir/test/partition_view_update_test.exs ==
** (CompileError) ../src/couchdb/test/elixir/test/partition_view_update_test.exs:1: cannot define module PartitionSizeTest because it is currently being defined in /Users/jay/proj/ibm/db/src/couchdb/test/elixir/test/partition_size_test.exs:1
    (stdlib) erl_eval.erl:670: :erl_eval.do_apply/6
    (elixir) lib/code.ex:677: Code.require_file/2
    (elixir) lib/kernel/parallel_compiler.ex:201: anonymous fn/4 in Kernel.ParallelCompiler.spawn_workers/6
make: *** [elixir-test] Error 1
returned 2
```

davisp replied:
Derp. Good catch. That replication test takes too long I was only ever running them in isolation.

tonysun83 commented:
When creating a mango json/text index, and the user doesn't specify`"ddoc" : "<partitionname>:<design_docid>"`, we get a 500 error (instead of a 400). 

```
{
    "error": "error_saving_ddoc",
    "reason": "Unknown error while saving the design document: unknown_error",
    "ref": 452476559
}
```
We know mango automatically creates the ddocid when not provided so it's not conforming to the partition format. I think the error message isn't handled correctly somewhere leading to a gen_server crash and  thus it doesn't get propagated up.


```
[info] 2019-01-07T06:06:47.415090Z node1@127.0.0.1 <0.24733.0> 9108cd090a 9108cd090a 127.0.0.1  localhost:5984 POST /partitiondb/_index 500 ok 8
[error] 2019-01-07T06:06:47.414666Z node3@127.0.0.1 <0.11851.0> -------- gen_server <0.11851.0> terminated with reason: bad return value {illegal_docid,<<"Doc id must be of form partition:id">>}
  last msg: {update_docs,<0.24906.0>,[[{doc,<<"_design/c1d29e81d94c805422e8b27d5891b97c0fa0a0d2">>,{1,[<<218,118,244,112,98,6,119,52,117,157,178,147,84,93,101,169>>]},[<<128,0,0,225>>,<<245,194,40,241,77,161,96,237,110,110,181,149,3,122,8,42>>,<<131,104,2,109,0,0,0,207,1,174,2,88,131,104,1,108,0,0,0,3,104,2,109,0,0,0,8,108,97,110,103,117,97,103,101,1,13,20,5,113,117,101,114,121,9,25,20,5,118,105,101,119,115,9,44,0,1,9,19,36,9,102,111,111,45,105,110,100,101,120,9,23,13,67,12,3,109,97,112,9,17,0,2,9,40,20,6,102,105,101,108,100,54,60,0,4,1,97,1,97,16,3,97,115,99,106,9,37,104,23,112,97,114,116,105,97,108,95,102,105,108,116,101,114,95,115,101,108,101,99,116,111,114,104,1,106,13,34,20,6,114,101,100,117,99,5,153,24,6,95,99,111,117,110,116,9,58,28,7,111,112,116,105,111,110,115,9,116,13,156,12,3,100,101,102,50,17,0,13,133,1,235,0,1,1,123,8,1,97,106,1,1,106,67,0,0,11,5,146,56,116,105,111,110,101,100,100,0,4,116,114,117,101,106,106,109,0,0,0,5,1,2,4,131,106>>],[],false,[{comp_body,<<1,174,2,88,131,104,1,108,0,0,0,3,104,2,109,0,0,0,8,108,97,110,103,117,97,103,101,1,13,20,5,113,117,101,114,121,9,25,20,5,118,105,101,119,115,9,44,0,1,9,19,36,9,102,111,111,45,105,110,100,101,120,9,23,13,67,12,3,109,97,112,9,17,0,2,9,40,20,6,102,105,101,108,100,54,60,0,4,1,97,1,97,16,3,97,115,99,106,9,37,104,23,112,97,114,116,105,97,108,95,102,105,108,116,101,114,95,115,101,108,101,99,116,111,114,104,1,106,13,34,20,6,114,101,100,117,99,5,153,24,6,95,99,111,117,110,116,9,58,28,7,111,112,116,105,111,110,115,9,116,13,156,12,3,100,101,102,50,17,0,13,133,1,235,0,1,1,123,8,1,97,106,1,1,106,67,0,0,11,5,146,56,116,105,111,110,101,100,100,0,4,116,114,117,101,106,106>>},{size_info,[]},{atts_stream,nil},{ejson_size,387},{ref,#Ref<0.2404497343.1436811265.145848>}]}]],[],false,true}
     state: {db,1,<<"shards/20000000-3fffffff/partitiondb.1546826084">>,"/Users/tonysun/github/cloudant/partitiondb/dev/lib/node3/data/dbs/shards/20000000-3fffffff/partitiondb.1546826084.couch",{couch_bt_engine,{st,"/Users/tonysun/github/cloudant/partitiondb/dev/lib/node3/data/dbs/shards/20000000-3fffffff/partitiondb.1546826084.couch",<0.11852.0>,#Ref<0.2404497343.1436811269.53299>,[before_header,after_header,on_file_open],{db_header,7,0,0,nil,nil,{12448,[],326},nil,nil,4253,1000,<<"d947ffdbbb88cfcceec1aa1dfa8ee453">>,[{'node3@127.0.0.1',0}],0,1000,0},false,{btree,<0.11852.0>,nil,#Fun<couch_bt_engine.id_tree_split.1>,#Fun<couch_bt_engine.id_tree_join.2>,undefined,#Fun<couch_bt_engine.id_tree_reduce.2>,snappy},{btree,<0.11852.0>,nil,#Fun<couch_bt_engine.seq_tree_split.1>,#Fun<couch_bt_engine.seq_tree_join.2>,undefined,#Fun<couch_bt_engine.seq_tree_reduce.2>,snappy},{btree,<0.11852.0>,{12448,[],326},#Fun<couch_bt_engine.local_tree_split.1>,#Fun<couch_bt_engine.local_tree_join.2>,undefined,nil,snappy},snappy,{btree,<0.11852.0>,nil,#Fun<couch_bt_engine.purge_tree_split.1>,#Fun<couch_bt_engine.purge_tree_join.2>,undefined,#Fun<couch_bt_engine.purge_tree_reduce.2>,snappy},{btree,<0.11852.0>,nil,#Fun<couch_bt_engine.purge_seq_tree_split.1>,#Fun<couch_bt_engine.purge_seq_tree_join.2>,undefined,#Fun<couch_bt_engine.purge_tree_reduce.2>,snappy}}},<0.11851.0>,nil,0,<<"1546840571318289">>,{user_ctx,null,[],undefined},[],undefined,nil,nil,nil,[{default_security_object,[]},{create_if_missing,true},{props,[{partitioned,true},{hash,[couch_partition,hash,[]]}]}],undefined}
```

tonysun83 commented:
After grabbing the latest commit on this branch. I no longer get an unknown error 500's. I think the code is working as designed. We might want to add in an extra field so users know which indexes have partitioned = false. 

```
{
    "total_rows": 4,
    "indexes": [
        {
            "ddoc": null,
            "name": "_all_docs",
            "type": "special",
            "def": {
                "fields": [
                    {
                        "_id": "asc"
                    }
                ]
            }
        },
        {
            "ddoc": "_design/foo-default-partition-index",
            "name": "foo-default-partition-index",
            "type": "json",
            "def": {
                "fields": [
                    {
                        "a": "asc"
                    }
                ]
            }
        },
        {
            "ddoc": "_design/foo-partition-false-index",
            "name": "foo-default-partition-false-index",
            "type": "json",
            "def": {
                "fields": [
                    {
                        "a": "asc"
                    }
                ]
            }
        },
        {
            "ddoc": "_design/foo-partition-true-index",
            "name": "foo-default-partition-true-index",
            "type": "json",
            "def": {
                "fields": [
                    {
                        "a": "asc"
                    }
                ]
            }
        }
    ]
}
```

```
{
    "_id": "_design/foo-default-partition-index",
    "_rev": "1-f55c91d8a173c5b1e7e61f35b3f4b594",
    "language": "query",
    "views": {
        "foo-default-partition-index": {
            "map": {
                "fields": {
                    "a": "asc"
                },
                "partial_filter_selector": {}
            },
            "reduce": "_count",
            "options": {
                "def": {
                    "fields": [
                        "a"
                    ]
                }
            }
        }
    }
}
```

tonysun83 commented:
we'll need to figure out what layer (core, dreyfus or clouseau) this issue is occuring, but it seems for text indexes, it still uses a `"partitioned": "false"` index for a partitoned query :

```
http://localhost:5984/partitiondb/_partition/foo/_explain

{
    "selector": {
      "a": 1
    },
  "use_index": "_design/11be2d55f04e03e19c5dc9a1a7038bb5a8a0f6f2"
}
```

```
{
    "dbname": "partitiondb",
    "index": {
        "ddoc": "_design/11be2d55f04e03e19c5dc9a1a7038bb5a8a0f6f2",
        "name": "11be2d55f04e03e19c5dc9a1a7038bb5a8a0f6f2",
        "type": "text",
        "def": {
            "default_analyzer": "keyword",
            "default_field": {},
            "selector": {},
            "fields": [],
            "index_array_lengths": true
        }
    },
    "partitioned": false,
    "selector": {
        "a": {
            "$eq": 1
        }
    },
    "opts": {
        "use_index": [
            "11be2d55f04e03e19c5dc9a1a7038bb5a8a0f6f2"
        ],
        "bookmark": [],
        "limit": 10000000000,
        "skip": 0,
        "sort": {},
        "fields": "all_fields",
        "partition": "foo",
        "r": [
            49
        ],
        "conflicts": false,
        "stale": false,
        "update": true,
        "stable": false,
        "execution_stats": false
    },
    "limit": 200,
    "skip": 0,
    "fields": "all_fields",
    "query": "(a_3anumber:1)",
    "partition": "foo",
    "sort": "relevance"
}
```

davisp commented:
@eiri Thanks for the review!

For the default yeah, we'll have to call that out in the docs that the paritioned flag defaults to the database flag.

For replications there's a create_target_params [1] option for replications where a user can specify the partitioned=true. I don't have an immediate good argument in either direction on whether we should default that to the partitioned option on the source db.

[1] https://github.com/apache/couchdb/commit/3e511b37bde8238918edc18c1bed9ab7ca1cbc5f

davisp commented:
@tonysun83 Good idea on adding the partitoned option. Have done that.

For that bug you found its because you're specifying an un-partitioned ddoc for a partition query and were weren't properly catching that after the fact. I've got a simple fix for that.

tonysun83 commented:
+1, new text index tests worked with partitioned dbs and mango text regression suite passed also

tonysun83 commented:
+1

AlexanderKaraberov commented:
Hello guys!
I'm trying to sync with this new cool feature as it totally passed by me. Please correct me if I'm wrong. As I understand this adds something akin to list partitioning in relational databases to CouchDB (on top of current range partitioning) where each partition is defined and selected based on the membership of a document id?
Could you please point me to other discussions/sources (mailing list perhaps?) where I can read more about this feature and ponder beforehand about the changes in our backend in order to leverage this feature. Thanks!

janl commented:
@AlexanderKaraberov https://lists.apache.org/thread.html/4d09ee0fcc4afec8f25dbd540f9c4f0af67d3fdcc41f26f36ff7a32d@%3Cdev.couchdb.apache.org%3E

AlexanderKaraberov commented:
@janl Thank you! Looks exactly like what I was looking for.
